pool/app/pool_common/entity_base_model.ml

```diff
@@ -0,0 +1,132 @@
+
+  let create m = m
+  let value m = m
+
+  let stringify = function
+    | true -> "true"
+    | false -> "false"
+  ;;
+
+  let of_string = function
+    | "true" -> true
+    | _ -> false
+  ;;
+
+  let schema field () =
+    Pool_common_utils.schema_decoder
+      (fun m ->
+        m
+        |> bool_of_string_opt
+        |> CCOption.get_or ~default:false
+        |> CCResult.pure)
+      string_of_bool
+      field
+  ;;
+end
+
+module type BooleanSig = sig
+  type t
+
+  val equal : t -> t -> bool
+  val pp : Format.formatter -> t -> unit
+  val show : t -> string
+  val sexp_of_t : t -> Ppx_sexp_conv_lib.Sexp.t
+  val t_of_yojson : Yojson.Safe.t -> t
+  val yojson_of_t : t -> Yojson.Safe.t
+  val create : bool -> t
+  val value : t -> bool
+  val stringify : t -> string
+  val of_string : string -> t
+
+  val schema
+    :  unit
+    -> (Entity_message.error, t) Pool_common_utils.PoolConformist.Field.t
+end
+
+module String = struct
+  open Sexplib.Conv
+
+  type t = string [@@deriving eq, show, sexp_of, yojson]
+
+  let value m = m
+
+  let create field str =
+    if CCString.is_empty str
+    then Error Entity_message.(Invalid field)
+    else Ok str
+  ;;
+
+  let schema field ?validation () =
+    let create = CCOption.value ~default:(create field) validation in
+    Pool_common_utils.schema_decoder create value field
+  ;;
+end
+
+module type StringSig = sig
+  type t
+
+  val equal : t -> t -> bool
+  val pp : Format.formatter -> t -> unit
+  val show : t -> string
+  val sexp_of_t : t -> Ppx_sexp_conv_lib.Sexp.t
+  val t_of_yojson : Yojson.Safe.t -> t
+  val yojson_of_t : t -> Yojson.Safe.t
+  val create : string -> (t, Entity_message.error) result
+  val value : t -> string
+
+  val schema
+    :  unit
+    -> (Entity_message.error, t) Pool_common_utils.PoolConformist.Field.t
+end
+
+module Integer = struct
+  open Sexplib.Conv
+
+  type t = int [@@deriving eq, show, sexp_of, yojson]
+
+  let value m = m
+
+  let schema field create () =
+    let decode str =
+      let open CCResult in
+      CCInt.of_string str
+      |> CCOption.to_result Entity_message.(NotANumber str)
+      >>= create
+    in
+    Pool_common_utils.schema_decoder decode CCInt.to_string field
+  ;;
+end
+
+module type IntegerSig = sig
+  type t
+
+  val equal : t -> t -> bool
+  val pp : Format.formatter -> t -> unit
+  val show : t -> string
+  val sexp_of_t : t -> Ppx_sexp_conv_lib.Sexp.t
+  val t_of_yojson : Yojson.Safe.t -> t
+  val yojson_of_t : t -> Yojson.Safe.t
+  val create : int -> (t, Entity_message.error) result
+  val value : t -> int
+
+  val schema
+    :  unit
+    -> (Entity_message.error, t) Pool_common_utils.PoolConformist.Field.t
+end
+
+module type BaseSig = sig
+  type t
+
+  val equal : t -> t -> bool
+  val pp : Format.formatter -> t -> unit
+  val show : t -> string
+  val create : string -> (t, Entity_message.error) result
+
+  val schema
+    :  unit
+    -> (Entity_message.error, t) Pool_common_utils.PoolConformist.Field.t
+end
```

Comment:
It might make sense to add another one for Ptime stuff? I've encountered it various times so far, what do you think?

Reply:
@aronerben 

Many modules have `type t = Ptime.t option`

Do you think it would make sense to rewrite those module to use 
```
type t = {
  canceled_at : CanceledAt.t option
}
```
instead of

```
module CanceledAt = struct
  type t = Ptime.t option [@@deriving eq, show]
end
```
Then we could reuse the same module more often.

Reply:
That is not was I meant :). I was talking about moving the option type out of the module on the record type.

Reply:
Ah now I understand :D Yea that change makes sense to me. It would make more sense in the domain as well since `CanceledAt` should only describe cancellation and not if it is optional or not.